### PR TITLE
Warn experimentalist when Ramsey freq change > artificial detuning

### PR DESCRIPTION
**Pull request by stephlazar**
Adds feature to warn the experimentalist if the qubit frequency change from a Ramsey experiment is larger than the chosen artificial detuning:

- some text in the Ramsey image text box will be highlighted in red
- an image containing a warning will be copied to the data folder
- can be tested with: a = tda.RamseyAnalysis(t_start='20210603_112017', options_dict={'artificial_detuning': 10e3})

### CODE REVIEW

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/timedomain_analysis.py]: Isn't the code easier as follows?
```
idx = [i for i, s in enumerate(textstr) if 'Delta f' in s][0]
```

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/timedomain_analysis.py]: Please add a comment that -1 is the line "artificial detuning" (or would it make sense to search for the index like you do for the other line?)

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/timedomain_analysis.py]: maybe you could replace "we don't want this" by a physics explanation

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/timedomain_analysis.py]: maybe self._raise_warning_image could automatically keep track of whether a warning has already been raised (e.g., by setting a self._warning_image_raised private property). Having the check there would make the code easier in all functions that call _raise_warning_image

File: pycqed/analysis/tools/plotting.py
Comment (chellings) [inline pycqed/analysis/tools/plotting.py:376]: please add a comment with the source when copying a function from the web

File: pycqed/analysis_v2/base_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/base_analysis.py]: As discussed orally: we might want to move the core functionality to a function in utilities. This function would require a destination folder. We could then have a wrapper here, which calls the utility function passing the folder of the current analysis. Then, in the future, other modules that know a timestamp folder (e.g., MC) could make use of the same utility function.

File: pycqed/analysis_v2/base_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/base_analysis.py]: While we might find the text on the waning image funny now, wouldn't a simpler warning sign (e.g., just the red warning text, or a yellow sign with exclamation mark) be a more sustainable solution?

Moreover, as discussed orally, it would make sense to log a warning message into a text file to inform the user what triggered the warning sign. (Note that you would need to keep appending to the text file if this function is called multiple times; even if you don't want to copy the warning image a second time.)

**Review by chellings — CHANGES_REQUESTED** [review]
Thanks, @stephlazar, for implementing this. As already discussed orally, I would have some suggestions. See my comments.

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/timedomain_analysis.py]: Yes, thanks

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/timedomain_analysis.py]: Yes, it makes sense to find index here too.

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/timedomain_analysis.py]: Indeed.

File: pycqed/analysis/tools/plotting.py
Comment (stephlazar) [reply pycqed/analysis/tools/plotting.py:376]: 👍

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/timedomain_analysis.py]: Makes sense 👍

File: pycqed/analysis_v2/base_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/base_analysis.py]: Makes sense, I moved it now to utilities.py.

File: pycqed/analysis_v2/base_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/base_analysis.py]: I added this functionality now.

File: pycqed/utilities/general.py
Comment (chellings) [inline pycqed/utilities/general.py:815]: Since you have the check here, you do not need the check with _warning_image_raised in ba

File: pycqed/utilities/general.py
Comment (chellings) [inline pycqed/utilities/general.py]: I would do this only if the filename does not have an extension yet

File: pycqed/utilities/general.py
Comment (chellings) [inline pycqed/utilities/general.py]: "ensure the if" -- > "ensure that if"

File: pycqed/utilities/general.py
Comment (chellings) [inline pycqed/utilities/general.py]: Why not? If a warning happens multiple times, I would expect to see it multiple times in the file (same behavior as the normal logging system)

**Review by chellings — APPROVED** [review]
Thanks for the updated version. I have added some final comments, and I would approve to merge after those are addressed.

File: pycqed/utilities/general.py
Comment (stephlazar) [reply pycqed/utilities/general.py:815]: Done

File: pycqed/utilities/general.py
Comment (stephlazar) [reply pycqed/utilities/general.py]: 👍 thanks

File: pycqed/utilities/general.py
Comment (stephlazar) [reply pycqed/utilities/general.py]: Changed now

File: pycqed/utilities/general.py
Comment (stephlazar) [reply pycqed/utilities/general.py]: My thought here was to prevent that the same message is appended each time we reanalyze the file offline. Do you want it to do this?

Comment (stephlazar) [reply pycqed/utilities/general.py]: Could you have another look at this before I merge? I now made some bigger changes.

This is what the warning file looks like now:
image

File: pycqed/analysis_v2/base_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/base_analysis.py]: docstring is outdated

File: pycqed/analysis_v2/base_analysis.py
Comment (chellings) [inline pycqed/analysis_v2/base_analysis.py:274]: Just some general comments (do not necessarily need to be addressed in this PR):
- Explaining an if statement with a comment that starts with the word "else" might cause confusion.
- I wonder what the best place for such an explanation is: in the docstring or here.
- I would prefer the following code in such an instance to avoid an additional level of indentation. It does not really matter in this special case since there are no deep indentation levels in this function, so it is, as I said, just a general comment
```
if self.check_plotting_delegation():
    return
```

**Review by chellings — APPROVED** [review]
Looks good to me. Just some minor comments, which you could consider before merging.

**Pull request merged**
